File: packages/dcflight/android/src/main/kotlin/com/dotcorr/dcflight/components/DCFScrollViewComponent.kt
/*
 * Copyright (c) Dotcorr Studio. and affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

package com.dotcorr.dcflight.components

import android.content.Context
import android.graphics.PointF
import android.os.Handler
import android.os.Looper
import android.util.Log
import android.view.View
import com.dotcorr.dcflight.components.DCFComponent
import com.dotcorr.dcflight.components.DCFNodeLayout
import com.dotcorr.dcflight.extensions.applyStyles

/**
 * DCFScrollViewComponent - Component manager
 * Manages creation and updates of DCFScrollView instances
 */
class DCFScrollViewComponent : DCFComponent() {
    
    companion object {
        private const val TAG = "DCFScrollViewComponent"
    }
    
    override fun createView(context: Context, props: Map<String, Any?>): View {
        val scrollView = DCFScrollView(context)
        
        // Don't create contentView here
        // Wait for ScrollContentView component to be attached via attachView
        // This prevents the assertion error when insertContentView is called twice
        
        // Basic setup
        scrollView.setShowsVerticalScrollIndicator(true)
        scrollView.setShowsHorizontalScrollIndicator(true)
        scrollView.setBounces(true)
        scrollView.setScrollEnabled(true)
        scrollView.setAlwaysBounceVertical(true)
        scrollView.setAlwaysBounceHorizontal(false)
        
        updateView(scrollView, props)
        val nonNullProps = props.filterValues { it != null }.mapValues { it.value!! }
        scrollView.applyStyles(nonNullProps)
        
        return scrollView
    }
    
    override fun updateView(view: View, props: Map<String, Any?>): Boolean {
        val scrollView = view as? DCFScrollView ?: return false
        
        // Scroll indicators
        (props["showsScrollIndicator"] as? Boolean)?.let { showsScrollIndicator ->
            scrollView.setShowsVerticalScrollIndicator(showsScrollIndicator)
            scrollView.setShowsHorizontalScrollIndicator(showsScrollIndicator)
        }
        
        // Bounces
        (props["bounces"] as? Boolean)?.let { bounces ->
            scrollView.setBounces(bounces)
        }
        
        // Horizontal scrolling
        (props["horizontal"] as? Boolean)?.let { horizontal ->
            if (horizontal) {
                scrollView.setAlwaysBounceHorizontal(true)
                scrollView.setAlwaysBounceVertical(false)
                scrollView.setShowsHorizontalScrollIndicator(scrollView.scrollView.isVerticalScrollBarEnabled)
                scrollView.setShowsVerticalScrollIndicator(false)
            } else {
                scrollView.setAlwaysBounceHorizontal(false)
                scrollView.setAlwaysBounceVertical(true)
                scrollView.setShowsVerticalScrollIndicator(false)
            }
        }
        
        // Paging
        (props["pagingEnabled"] as? Boolean)?.let { pagingEnabled ->
            val contentOffset = PointF(scrollView.scrollView.scrollX.toFloat(), scrollView.scrollView.scrollY.toFloat())
            // Android doesn't have isPagingEnabled, handled via custom logic if needed
            scrollView.scrollView.scrollTo(contentOffset.x.toInt(), contentOffset.y.toInt())
        }
        
        // Scroll enabled
        (props["scrollEnabled"] as? Boolean)?.let { scrollEnabled ->
            scrollView.setScrollEnabled(scrollEnabled)
        }
        
        // Content insets
        (props["contentInset"] as? Map<*, *>)?.let { contentInset ->
            val top = (contentInset["top"] as? Number)?.toFloat()?.toInt() ?: 0
            val left = (contentInset["left"] as? Number)?.toFloat()?.toInt() ?: 0
            val bottom = (contentInset["bottom"] as? Number)?.toFloat()?.toInt() ?: 0
            val right = (contentInset["right"] as? Number)?.toFloat()?.toInt() ?: 0
            scrollView.contentInset = android.graphics.Rect(left, top, right, bottom)
        }
        
        // Center content
        (props["centerContent"] as? Boolean)?.let { centerContent ->
            scrollView.centerContent = centerContent
        }
        
        // Scroll event throttle
        (props["scrollEventThrottle"] as? Number)?.toDouble()?.let { throttle ->
            scrollView.scrollEventThrottle = throttle.toLong()
        }
        
        // Handle commands
        handleCommand(scrollView, props)
        
        val nonNullProps = props.filterValues { it != null }.mapValues { it.value!! }
        scrollView.applyStyles(nonNullProps)
        return true
    }
    
    override fun applyLayout(view: View, layout: DCFNodeLayout) {
        val scrollView = view as? DCFScrollView ?: return
        
        // Apply layout to scroll view
        scrollView.layout(
            layout.left.toInt(),
            layout.top.toInt(),
            (layout.left + layout.width).toInt(),
            (layout.top + layout.height).toInt()
        )
        
        // CRITICAL: Don't update contentSize here because ScrollContentView's layout hasn't been applied yet
        // Layouts are applied parent-first, so ScrollView's layout is applied before ScrollContentView's layout
        // ScrollContentViewComponent.applyLayout will trigger updateContentSizeFromContentView after it sets its frame
        // This ensures we read the correct frame size
        Log.d(TAG, "🔍 DCFScrollViewComponent.applyLayout: Applied frame=(${scrollView.left}, ${scrollView.top}, ${scrollView.width}, ${scrollView.height}), deferring contentSize update until ScrollContentView layout is applied")
    }
    
    override fun viewRegisteredWithShadowTree(view: View, shadowNode: com.dotcorr.dcflight.layout.DCFShadowNode, nodeId: String) {
        val scrollView = view as? DCFScrollView ?: return
        
        view.setTag("nodeId".hashCode(), nodeId)
        
        Handler(Looper.getMainLooper()).post {
            scrollView.updateContentSizeFromContentView()
        }
    }
    
    override fun handleTunnelMethod(method: String, arguments: Map<String, Any?>): Any? {
        return null
    }
    
    /// Handle commands
    private fun handleCommand(scrollView: DCFScrollView, props: Map<String, Any?>) {
        val commandData = props["command"] as? Map<*, *> ?: return
        
        (commandData["scrollToPosition"] as? Map<*, *>)?.let { scrollToPositionData ->
            val x = (scrollToPositionData["x"] as? Number)?.toDouble()
            val y = (scrollToPositionData["y"] as? Number)?.toDouble()
            if (x != null && y != null) {
                val animated = (scrollToPositionData["animated"] as? Boolean) ?: true
                scrollView.scrollToOffset(PointF(x.toFloat(), y.toFloat()), animated)
            }
        }
        
        (commandData["scrollToTop"] as? Map<*, *>)?.let { scrollToTopData ->
            val animated = (scrollToTopData["animated"] as? Boolean) ?: true
            scrollView.scrollToOffset(PointF(scrollView.scrollView.scrollX.toFloat(), 0f), animated)
        }
        
        (commandData["scrollToBottom"] as? Map<*, *>)?.let { scrollToBottomData ->
            val animated = (scrollToBottomData["animated"] as? Boolean) ?: true
            scrollView.scrollToEnd(animated)
        }
        
        (commandData["flashScrollIndicators"] as? Boolean)?.let { flashScrollIndicators ->
            if (flashScrollIndicators) {
                // Android doesn't have flashScrollIndicators, handled via custom logic if needed
            }
        }
    }
}

File: packages/dcflight/ios/Classes/Coordination/Components/DCFScrollViewComponent.swift
/*
 * Copyright (c) Dotcorr Studio. and affiliates.
 *
 * Licensed under the PolyForm Noncommercial License 1.0.0.
 * Commercial use requires a license from DotCorr.
 */

import UIKit
import dcflight

/**
 * DCFScrollViewComponent - Component manager
 * Manages creation and updates of DCFScrollView instances
 */
class DCFScrollViewComponent: NSObject, DCFComponent {
    required override init() {
        super.init()
    }
    
    func createView(props: [String: Any]) -> UIView {
        let scrollView = DCFScrollView(frame: .zero)
        
        // Don't create contentView here
        // Wait for ScrollContentView component to be attached via attachView
        // This prevents the assertion error when insertContentView is called twice
        
        // Basic setup
        scrollView.scrollView.showsVerticalScrollIndicator = true
        scrollView.scrollView.showsHorizontalScrollIndicator = true
        scrollView.scrollView.bounces = true
        scrollView.scrollView.isScrollEnabled = true
        scrollView.setAlwaysBounceVertical(true)
        scrollView.setAlwaysBounceHorizontal(false)
        
        if #available(iOS 11.0, *) {
            scrollView.scrollView.contentInsetAdjustmentBehavior = .never
        }
        
        updateView(scrollView, withProps: props)
        scrollView.applyStyles(props: props)
        
        return scrollView
    }
    
    func updateView(_ view: UIView, withProps props: [String: Any]) -> Bool {
        guard let scrollView = view as? DCFScrollView else { return false }
        
        // Scroll indicators
        if let showsScrollIndicator = props["showsScrollIndicator"] as? Bool {
            scrollView.setShowsVerticalScrollIndicator(showsScrollIndicator)
            scrollView.setShowsHorizontalScrollIndicator(showsScrollIndicator)
        }
        
        // Bounces
        if let bounces = props["bounces"] as? Bool {
            scrollView.setBounces(bounces)
        }
        
        // Horizontal scrolling
        if let horizontal = props["horizontal"] as? Bool {
            if horizontal {
                scrollView.setAlwaysBounceHorizontal(true)
                scrollView.setAlwaysBounceVertical(false)
                scrollView.setShowsHorizontalScrollIndicator(scrollView.scrollView.showsVerticalScrollIndicator)
                scrollView.setShowsVerticalScrollIndicator(false)
            } else {
                scrollView.setAlwaysBounceHorizontal(false)
                scrollView.setAlwaysBounceVertical(true)
                scrollView.setShowsHorizontalScrollIndicator(false)
            }
        }
        
        // Paging
        if let pagingEnabled = props["pagingEnabled"] as? Bool {
            let contentOffset = scrollView.scrollView.contentOffset
            scrollView.scrollView.isPagingEnabled = pagingEnabled
            scrollView.scrollView.contentOffset = contentOffset
        }
        
        // Scroll enabled
        if let scrollEnabled = props["scrollEnabled"] as? Bool {
            scrollView.setScrollEnabled(scrollEnabled)
        }
        
        // Content insets
        if let contentInset = props["contentInset"] as? [String: Any] {
            let top = (contentInset["top"] as? CGFloat) ?? 0
            let left = (contentInset["left"] as? CGFloat) ?? 0
            let bottom = (contentInset["bottom"] as? CGFloat) ?? 0
            let right = (contentInset["right"] as? CGFloat) ?? 0
            scrollView.contentInset = UIEdgeInsets(top: top, left: left, bottom: bottom, right: right)
        }
        
        // Center content
        if let centerContent = props["centerContent"] as? Bool {
            scrollView.centerContent = centerContent
        }
        
        // Scroll event throttle
        if let throttle = props["scrollEventThrottle"] as? Double {
            scrollView.scrollEventThrottle = throttle
        }
        
        // Handle commands
        handleCommand(scrollView: scrollView, props: props)
        
        scrollView.applyStyles(props: props)
        return true
    }
    
    func applyLayout(_ view: UIView, layout: YGNodeLayout) {
        guard let scrollView = view as? DCFScrollView else { return }
        
        // Apply layout to scroll view
        scrollView.frame = CGRect(x: layout.left, y: layout.top, width: layout.width, height: layout.height)
        
        // CRITICAL: Don't update contentSize here because ScrollContentView's layout hasn't been applied yet
        // Layouts are applied parent-first, so ScrollView's layout is applied before ScrollContentView's layout
        // ScrollContentViewComponent.applyLayout will trigger updateContentSizeFromContentView after it sets its frame
        // This ensures we read the correct frame size
        print("🔍 DCFScrollViewComponent.applyLayout: Applied frame=\(scrollView.frame), deferring contentSize update until ScrollContentView layout is applied")
    }
    
    func viewRegisteredWithShadowTree(_ view: UIView, shadowView: DCFShadowView, nodeId: String) {
        guard let scrollView = view as? DCFScrollView else { return }
        
        objc_setAssociatedObject(view,
                                 UnsafeRawPointer(bitPattern: "nodeId".hashValue)!,
                                 nodeId,
                                 .OBJC_ASSOCIATION_RETAIN_NONATOMIC)
        
        // Update content size after children are added
        DispatchQueue.main.async {
            scrollView.updateContentSizeFromContentView()
        }
    }
    
    
    /// Handle commands
    private func handleCommand(scrollView: DCFScrollView, props: [String: Any]) {
        guard let commandData = props["command"] as? [String: Any] else {
            return
        }
        
        if let scrollToPositionData = commandData["scrollToPosition"] as? [String: Any] {
            if let x = scrollToPositionData["x"] as? Double, let y = scrollToPositionData["y"] as? Double {
                let animated = scrollToPositionData["animated"] as? Bool ?? true
                scrollView.scrollToOffset(CGPoint(x: CGFloat(x), y: CGFloat(y)), animated: animated)
            }
        }
        
        if let scrollToTopData = commandData["scrollToTop"] as? [String: Any] {
            let animated = scrollToTopData["animated"] as? Bool ?? true
            scrollView.scrollToOffset(CGPoint(x: scrollView.scrollView.contentOffset.x, y: 0), animated: animated)
        }
        
        if let scrollToBottomData = commandData["scrollToBottom"] as? [String: Any] {
            let animated = scrollToBottomData["animated"] as? Bool ?? true
            scrollView.scrollToEnd(animated)
        }
        
        if let flashScrollIndicators = commandData["flashScrollIndicators"] as? Bool, flashScrollIndicators {
            scrollView.scrollView.flashScrollIndicators()
        }
    }
    
    static func handleTunnelMethod(_ method: String, params: [String: Any]) -> Any? {
        return nil
    }
}

